File: skills/agents-orchestrator/scripts/execution_secrets.py
"""Run-scoped child token seeds and deterministic Attempt token derivation."""

import base64
import hashlib
import hmac
import os
import pathlib

import state_store


SEED_BYTES = 32


def _secret_root():
    root = state_store.runtime_root() / "secrets"
    root.mkdir(parents=True, exist_ok=True, mode=0o700)
    os.chmod(str(root), 0o700)
    return root.resolve()


def seed_ref(root_id):
    return "secrets/%s.key" % root_id


def resolve_seed_path(reference):
    if not isinstance(reference, str) or not reference:
        raise RuntimeError("Run child token seed reference is missing")
    relative = pathlib.PurePosixPath(reference)
    if relative.is_absolute() or ".." in relative.parts or relative.parts[:1] != ("secrets",):
        raise RuntimeError("Run child token seed reference is invalid")
    path = (state_store.runtime_root() / pathlib.Path(*relative.parts)).resolve()
    if path.parent != _secret_root():
        raise RuntimeError("Run child token seed escapes the secret directory")
    return path


def create_run_seed(root_id):
    reference = seed_ref(root_id)
    path = resolve_seed_path(reference)
    seed = os.urandom(SEED_BYTES)
    flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
    if hasattr(os, "O_NOFOLLOW"):
        flags |= os.O_NOFOLLOW
    descriptor = os.open(str(path), flags, 0o600)
    try:
        os.write(descriptor, seed)
        os.fsync(descriptor)
    finally:
        os.close(descriptor)
    os.chmod(str(path), 0o600)
    return reference, hashlib.sha256(seed).hexdigest()


def _read_seed(run):
    path = resolve_seed_path(run.get("token_seed_ref"))
    try:
        mode = path.stat().st_mode & 0o777
        if mode != 0o600:
            raise RuntimeError("Run child token seed must have mode 0600")
        seed = path.read_bytes()
    except OSError as exc:
        raise RuntimeError("Run child token seed is unavailable") from exc
    if len(seed) != SEED_BYTES:
        raise RuntimeError("Run child token seed has an invalid length")
    expected = run.get("token_seed_hash")
    if not expected or not hmac.compare_digest(hashlib.sha256(seed).hexdigest(), expected):
        raise RuntimeError("Run child token seed hash does not match")
    return seed


def derive_attempt_token(run, attempt_id):
    seed = _read_seed(run)
    message = "%s|%s" % (run["root_id"], attempt_id)
    digest = hmac.new(seed, message.encode("utf-8"), hashlib.sha256).digest()
    return base64.urlsafe_b64encode(digest).decode("ascii").rstrip("=")


def remove_run_seed(run):
    reference = run.get("token_seed_ref") if run else None
    if not reference:
        return True
    path = resolve_seed_path(reference)
    try:
        path.unlink()
    except FileNotFoundError:
        pass
    return not path.exists()


def cleanup_run_seed_if_safe(root_id):
    with state_store.transaction(immediate=False) as con:
        run = state_store.get_run(root_id, con)
        if run is None or run["status"] not in {"done", "failed", "cancelled"}:
            return False
        open_effects = con.execute(
            """SELECT COUNT(*) AS n FROM effects
               WHERE root_id=? AND effect_type IN ('spawn_agent','stop_agent')
                 AND status IN ('pending','running')""",
            (root_id,),
        ).fetchone()["n"]
        open_executions = con.execute(
            """SELECT COUNT(*) AS n FROM launches l
               JOIN attempts a ON a.attempt_id=l.attempt_id
               JOIN tasks t ON t.task_id=a.task_id
               WHERE t.root_id=? AND l.status != 'closed'""",
            (root_id,),
        ).fetchone()["n"]
    if open_effects or open_executions:
        return False
    return remove_run_seed(run)

File: skills/agents-orchestrator/scripts/execution_secrets.ts
import {
  chmodSync,
  closeSync,
  constants as fsConstants,
  existsSync,
  fsyncSync,
  mkdirSync,
  openSync,
  readFileSync,
  statSync,
  unlinkSync,
  writeSync,
} from "node:fs";
import { createHash, createHmac, randomBytes, timingSafeEqual } from "node:crypto";
import { dirname, isAbsolute, join, resolve } from "node:path";

import * as stateStore from "./state_store.ts";
import { RuntimeError, type RuntimeRecord } from "./runtime_types.ts";

export const SEED_BYTES = 32;

function secretRoot(): string {
  const root = join(stateStore.runtimeRoot(), "secrets");
  mkdirSync(root, { recursive: true, mode: 0o700 });
  chmodSync(root, 0o700);
  return resolve(root);
}

export function seedRef(rootId: string): string {
  return `secrets/${rootId}.key`;
}

export function resolveSeedPath(reference: unknown): string {
  if (typeof reference !== "string" || !reference) {
    throw new RuntimeError("Run child token seed reference is missing");
  }
  const parts = reference.split("/");
  if (isAbsolute(reference) || parts.includes("..") || parts[0] !== "secrets" || parts.length !== 2) {
    throw new RuntimeError("Run child token seed reference is invalid");
  }
  const path = resolve(stateStore.runtimeRoot(), ...parts);
  if (dirname(path) !== secretRoot()) throw new RuntimeError("Run child token seed escapes the secret directory");
  return path;
}

export function createRunSeed(rootId: string): [string, string] {
  const reference = seedRef(rootId);
  const path = resolveSeedPath(reference);
  const seed = randomBytes(SEED_BYTES);
  const noFollow = "O_NOFOLLOW" in fsConstants ? fsConstants.O_NOFOLLOW : 0;
  const descriptor = openSync(path, fsConstants.O_WRONLY | fsConstants.O_CREAT | fsConstants.O_EXCL | noFollow, 0o600);
  try {
    writeSync(descriptor, seed);
    fsyncSync(descriptor);
  } finally {
    closeSync(descriptor);
  }
  chmodSync(path, 0o600);
  return [reference, createHash("sha256").update(seed).digest("hex")];
}

function readSeed(run: RuntimeRecord): Buffer {
  const path = resolveSeedPath(run.token_seed_ref);
  let seed: Buffer;
  try {
    if ((statSync(path).mode & 0o777) !== 0o600) {
      throw new RuntimeError("Run child token seed must have mode 0600");
    }
    seed = readFileSync(path);
  } catch (error) {
    if (error instanceof RuntimeError) throw error;
    throw new RuntimeError("Run child token seed is unavailable");
  }
  if (seed.byteLength !== SEED_BYTES) throw new RuntimeError("Run child token seed has an invalid length");
  const expected = run.token_seed_hash;
  const actual = createHash("sha256").update(seed).digest();
  if (typeof expected !== "string") throw new RuntimeError("Run child token seed hash does not match");
  const expectedBytes = Buffer.from(expected, "hex");
  if (expectedBytes.byteLength !== actual.byteLength || !timingSafeEqual(actual, expectedBytes)) {
    throw new RuntimeError("Run child token seed hash does not match");
  }
  return seed;
}

export function deriveAttemptToken(run: RuntimeRecord, attemptId: number): string {
  const digest = createHmac("sha256", readSeed(run))
    .update(`${run.root_id}|${attemptId}`, "utf8")
    .digest();
  return digest.toString("base64url");
}

export function removeRunSeed(run: RuntimeRecord | null): boolean {
  const reference = run?.token_seed_ref;
  if (!reference) return true;
  const path = resolveSeedPath(reference);
  try {
    unlinkSync(path);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
  return !existsSync(path);
}

export function cleanupRunSeedIfSafe(rootId: string): boolean {
  const result = stateStore.transaction((connection) => {
    const run = stateStore.getRun(rootId, connection);
    if (run === null || !new Set(["done", "failed", "cancelled"]).has(run.status)) {
      return { safe: false, run };
    }
    const openEffects = Number(connection.execute(
      `SELECT COUNT(*) AS n FROM effects
        WHERE root_id=? AND effect_type IN ('spawn_agent','stop_agent')
          AND status IN ('pending','running')`,
      [rootId],
    ).fetchone()?.n ?? 0);
    const openExecutions = Number(connection.execute(
      `SELECT COUNT(*) AS n FROM launches l
        JOIN attempts a ON a.attempt_id=l.attempt_id
        JOIN tasks t ON t.task_id=a.task_id
       WHERE t.root_id=? AND l.status != 'closed'`,
      [rootId],
    ).fetchone()?.n ?? 0);
    return { safe: openEffects === 0 && openExecutions === 0, run };
  }, false);
  return result.safe && result.run !== null ? removeRunSeed(result.run) : false;
}
